test-case/latency-metrics.sh

```diff
@@ -0,0 +1,156 @@
+##
+## Description:
+##    Run `jackd` audio server; execute `jack_iodelay` with its in/out ports
+##    connected to the loopback-ed ports and give it some time ot collect
+##    latency measurements (on each 0.5 sec.)
+##
+## Case step:
+##    1. Probe to start `jackd` with parameters given and read configuration.
+##    2. Start `jackd` again for latency measurements.
+##    3. Start `jack_iodelay` which awaits for its ports connected to loopback.
+##    4. Connect `jack_iodelay` ports to signal path ports with the loopback.
+##    5. Wait for the period given to collect enough latency measurements.
+##    6. Calculate latency statistics and save into a JSON file.
+##
+## Expect result:
+##    Latency statistics collected and saved in `results.json`.
+##    Exit status 0.
+##
+
+TESTDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+TESTLIB="${TESTDIR}/case-lib"
+
+# shellcheck source=case-lib/lib.sh
+source "${TESTLIB}/lib.sh"
+
+OPT_NAME['R']='run_period'      OPT_DESC['R']='Time period (in seconds) to measure latency.'
+OPT_HAS_ARG['R']=1              OPT_VAL['R']="30"
+
+OPT_NAME['d']='device'          OPT_DESC['d']='ALSA pcm device to use by JACK'
+OPT_HAS_ARG['d']=1              OPT_VAL['d']="hw:0"
+
+OPT_NAME['r']='rate'            OPT_DESC['r']='Sample rate to try latency with'
+OPT_HAS_ARG['r']=1              OPT_VAL['r']=48000
+
+OPT_NAME['S']='shorts'          OPT_DESC['S']='Try to use 16-bit samples instead of 32-bit, if possible.'
+OPT_HAS_ARG['S']=0              OPT_VAL['S']=0
+
+OPT_NAME['p']='port_p'          OPT_DESC['p']='Jack playback port with loopback. Example: system:playback_1'
+OPT_HAS_ARG['p']=1              OPT_VAL['p']=''
+
+OPT_NAME['c']='port_c'          OPT_DESC['c']='Jack capture port with loopback. Example: system:capture_1'
+OPT_HAS_ARG['c']=1              OPT_VAL['c']=''
+
+OPT_NAME['s']='sof-logger'      OPT_DESC['s']="Open sof-logger trace the data will store at $LOG_ROOT"
+OPT_HAS_ARG['s']=0              OPT_VAL['s']=1
+
+OPT_NAME['v']='verbose'         OPT_DESC['v']='Verbose logging.'
+OPT_HAS_ARG['v']=0              OPT_VAL['v']=0
+
+func_opt_parse_option "$@"
+
+alsa_device=${OPT_VAL['d']}
+alsa_shorts=$([ "${OPT_VAL['S']}" -eq 1 ] && echo '--shorts' || echo '')
+port_playback=${OPT_VAL['p']}
+port_capture=${OPT_VAL['c']}
+rate=${OPT_VAL['r']}
+run_period=${OPT_VAL['R']}
+run_verbose=$([ "${OPT_VAL['v']}" -eq 1 ] && echo '--verbose' || echo '')
+
+RUN_PERIOD_MAX="$((run_period + 30))s"
+JACKD_TIMEOUT="$((run_period + 15))s"
+JACKD_OPTIONS="${run_verbose} --realtime --temporary"
+JACKD_BACKEND="alsa"
+JACKD_BACKEND_OPTIONS="-d ${alsa_device} -r ${rate} ${alsa_shorts}"
+WAIT_JACKD="2s"
+IODELAY_TIMEOUT="${run_period}s"
+WAIT_IODELAY="2s"
+
+if [ "$port_playback" == "" ] || [ "$port_capture" == "" ];
+then
+  skip_test "No playback or capture Jack port is specified. Skip the test."
+fi
+
+setup_kernel_check_point
+
+start_test
+
+logger_disabled || func_lib_start_log_collect
```

Comment:
I know it hasn't been the usual practice, but I think these should be moved to `main()` too. Once they are moved to main(), you can very quickly comment out the final call to `main()` and source the entire file interactively to test any function interactively. For instance you can invoke `options2json` only.

Reply:
BTW the `main()` definition should IMHO be near the top, with only the `main "$@"` call at the bottom. This hasn't been the case in sof-test so far but this brand new test is a great opportunity to set a new standard :-)